h2o-py/h2o/explanation/_explain.py

```diff
@@ -1305,35 +1308,59 @@ def ice_plot(
             factor_map = _factor_mapper(NumpyFrame(frame[column]).from_factor_to_num(column))
 
         deciles = [int(round(frame.nrow * dec / 10)) for dec in range(11)]
+        deciles[10] = frame.nrow - 1
         colors = plt.get_cmap(colormap, 11)(list(range(11)))
         plt.figure(figsize=figsize)
         for i, index in enumerate(deciles):
+            percentile_string = "{}th Percentile".format(i * 10)
             tmp = NumpyFrame(
                 model.partial_plot(
                     frame,
                     cols=[column],
                     plot=False,
                     row_index=index,
                     targets=target,
-                    nbins=20 if not is_factor else 1 + frame[column].nlevels()[0]
+                    nbins=100 if not is_factor else 1 + frame[column].nlevels()[0],
+                    include_na=True
                 )[0]
             )
+            user_splits = dict()
+            orig_value = frame.as_data_frame()[column][index]
             encoded_col = tmp.columns[0]
+            if _isnan(orig_value):
+                id = np.where(tmp[encoded_col] == tmp.from_factor_to_num(encoded_col)['.missing(NA)'])[0][0] if is_factor else np.where(np.isnan(tmp[encoded_col]))[0][0]
```

Comment:
nit/style: I'd tend to avoid usage of `id` name, as it's built-in python function

nit/style: I find really hard to understand what e.g. 
```pythonnp.where(tmp[encoded_col] == tmp.from_factor_to_num(encoded_col)['.missing(NA)'])[0][0]```
 is doing. Is `".missing(NA)"` some convention for naming the missing values? If so, maybe it would be worth defining it as a constant. Also what happens if there is no such value found, it seems to fail on `IndexError` or so?

Reply:
yesyes, ".missing(NA)" is a convention for naming missing values in `tmp` in case of factor column. I can put it into constant.

The thing behind this is, that I am looking for the original value to plot it. It can either be a non-missing value or a missing value and I need to calculate respective response value.

- In case it is a non-missing value, I calculate partial_plot with a given `index `and predefined `user_splits`. into `user_splits` I put only the x-axis value of the original value so that I get response value at this point calculated by exactly the same transformations as points of ice line. Previous call with bins calculated points of the ice line - as many points as many bins, this call calculates only one (the original) point, with x-value defined in user_splits.

under the hood it does this:
```
try {
  preds = _model_id.get().score(test, Key.make().toString(), _job, false);
  if (preds == null || preds.numRows() == 0) {  // this can happen if algo will not predict on rows with NAs
    _meanResponse[_pdp_row_index] = Double.NaN;
    _stddevResponse[_pdp_row_index] = Double.NaN;
    _stdErrorOfTheMeanResponse[_pdp_row_index] = Double.NaN;
  } else {
    CalculateWeightMeanSTD calMeansSTD = (_weight_column_index >= 0)?
            getWeightedStat(fr, preds, _predictorColumn):null;
    _meanResponse[_pdp_row_index] = (_weight_column_index >= 0)?calMeansSTD.getWeightedMean()
            :preds.vec(_predictorColumn).mean();
    _stddevResponse[_pdp_row_index] = (_weight_column_index >= 0)?calMeansSTD.getWeightedSigma()
            :preds.vec(_predictorColumn).sigma();
    _stdErrorOfTheMeanResponse[_pdp_row_index] = _stddevResponse[_pdp_row_index]/Math.sqrt(preds.numRows());
  }
```
in PartialDependence.java

if I only used model.score() value for the original value calculation, for example for the binary model I would need to solve cases what to use as a response from scoredd output which looks like this

`predict        p0        p1
1       1 0.3597337 0.6402663`
 and I would need to solve weights etc to everythiing be in accordance with what is already displayed. By calling partial dependence on custom point defined by 'user_splits' I get a value which is in accordance with other ice plot points and I can be sure it is in the same scale with the rest of the plot and gone through the same transformations.


-  In case it is a missing value, I also need to calculate this value separately, but in h2o it is not possible to pass the nan to  'user_splits'. Hence, I calculate nan value right with the other ice line points by adding  `include_na=True`. Based on type of 'column' it is not guaranteed, which line in tmp will contain the result for nan. So I need to find that index and that's what line 
`idx = np.where(tmp[encoded_col] == tmp.from_factor_to_num(encoded_col)['.missing(NA)'])[0][0] if is_factor else np.where(np.isnan(tmp[encoded_col]))[0][0]` does. The value will be there, because I specified by include_na=True to be calculated, so IndexError can't happen and if it happens, then something got broken in partial dependence calculation and the calculation is corrupted anyway.
Next line 
`tmp.from_num_to_factor(encoded_col)[tmp[encoded_col][idx]] if is_factor else tmp[encoded_col][idx]`
gives me its x-value in tmp, I could use `orig_null_value = '.missing(NA)' if is_factor else np.nan` instead of it. It served me as a proof of correctness of selected row in tmp, will leave it commented there in case some bug tracing will be needed.

Reply:
[77c3d0a](https://github.com/h2oai/h2o-3/pull/6020/commits/77c3d0acca2ed68f6c134d67015734bc7c5af0f3)

Reply:
Hmm I have to admit that this is really confusing and if I was debugging this code later I probably wouldn't have figured out 😓 I expect you agreed on that beforehand, or maybe it's part of the requirements spec idk, but is it really needed to display original points for missing value then? Maybe I'm missing something, but the point itself is then just an interpolation, which hypothetically can be incorrect. Anyways, I don't want to dig into it too much, so it's fine from my perspective then. Thank you for explanation!

Reply:
It's good that it's discussed 👍  It's not an interpolation. For certain original value I am calculating pdp/ice in one exact point, the result of `orig_tmp = NumpyFrame( model.partial_plot(... user_splits = [orig_value] ...))` is one exact value to paint a point, not multiple values to be connected into a line. In line `preds = _model_id.get().score(test, Key.make().toString(), _job, false);` from the snippet I've sent above `test` frame consist of  1 row -> `preds `consist of  1 row -> `preds.vec(_predictorColumn).mean() `is mean out of 1 value -> `orig_tmp` consist of 1 row -> 1 point to paint, not multiple points to be connected into a line.

The another part of rquirement is to be able to plot not only NAs, but also user-specified invalid values, so the plot will be supposed to have "numerical part" of x-axis and "NA/invalid part" of x-axis // or there will be two plots, one with valid data, second with NA/invalid data depending on what will be doable. But this part will be done in https://h2oai.atlassian.net/browse/PUBDEV-8493.

<img width="618" alt="Screenshot 2022-02-11 at 18 50 44" src="https://user-images.githubusercontent.com/18303397/153643085-912159d9-5372-4503-9b72-1c2fb8b2b513.png">